FirebaseDatabaseUI/FirebaseIndexArray.h

```diff
@@ -0,0 +1,175 @@
+
+/**
+ * Delegate method which is called whenever an object is changed in a
+ * FirebaseArray. On a FirebaseArray synchronized to a Firebase reference, 
+ * this corresponds to a @c FIRDataEventTypeChildChanged event being raised.
+ * @param ref The database reference that changed in the array
+ * @param index The index the reference was changed at
+ */
+- (void)array:(FirebaseIndexArray *)array didChangeReference:(FIRDatabaseReference *)ref atIndex:(NSUInteger)index;
+
+/**
+ * Delegate method which is called whenever an object is removed from a
+ * FirebaseArray. On a FirebaseArray synchronized to a Firebase reference, 
+ * this corresponds to a @c FIRDataEventTypeChildRemoved event being raised.
+ * @param ref The database reference removed from the array
+ * @param index The index the reference was removed at
+ */
+- (void)array:(FirebaseIndexArray *)array didRemoveReference:(FIRDatabaseReference *)ref atIndex:(NSUInteger)index;
+
+/**
+ * Delegate method which is called whenever an object is moved within a
+ * FirebaseArray. On a FirebaseArray synchronized to a Firebase reference, 
+ * this corresponds to a @c FIRDataEventTypeChildMoved event being raised.
+ * @param ref The database reference that has moved locations
+ * @param fromIndex The index the reference is being moved from
+ * @param toIndex The index the reference is being moved to
+ */
+- (void)array:(FirebaseIndexArray *)array didMoveReference:(FIRDatabaseReference *)ref fromIndex:(NSUInteger)fromIndex toIndex:(NSUInteger)toIndex;
+
+/**
+ * Delegate method which is called whenever the backing query is canceled. This error is fatal
+ * and the index array will become unusable afterward, so please handle it appropriately
+ * (i.e. by displaying a modal error explaining why there's no content).
+ * @param error the error that was raised
+ */
+- (void)array:(FirebaseIndexArray *)array queryCancelledWithError:(NSError *)error;
+
+@end
+
+/**
+ * A FirebaseIndexArray instance uses a query's contents to query children of
+ * a separate database reference, which is useful for displaying an indexed list
+ * of data as described in https://firebase.google.com/docs/database/ios/structure-data
+ */
+@interface FirebaseIndexArray : NSObject
+
+/**
+ * An immutable copy of the loaded contents in the array. Returns an
+ * empty array if no contents have loaded yet.
+ */
+@property(nonatomic, copy, readonly) NSArray<FIRDataSnapshot *> *items;
+
+/**
+ * The delegate that this array should forward events to.
+ */
+@property(nonatomic, weak) id<FirebaseIndexArrayDelegate> delegate;
+
+/**
+ * Returns the number of items in the array.
+ */
+@property(nonatomic, readonly) NSUInteger count;
+
+- (instancetype)init NS_UNAVAILABLE;
+
+/**
+ * Initializes a FirebaseIndexArray with an index query and a data query.
+ * The array expects the keys of the children of the index query to be children
```

Comment:
(here and elsewhere) change "to be" to "to match" - this is complex stuff, and clarifying that we have the same values in two places is probably pertinent.